File: tiffany-frontend/src/redux/features/business/store/models/request/pos-checkout-request.ts
/**
 * POS Checkout Request Models
 * Matches backend POSCheckoutRequest and nested DTOs exactly.
 */

import { ItemPricingSnapshot, OrderPricingSnapshot } from "../type/pos-page-type";

// Item in the cart — matches backend POSCheckoutItemRequest
export interface POSCheckoutItemRequest {
  productId: string;
  productSizeId?: string | null;
  quantity: number;

  // Display fields
  productName?: string;
  productImageUrl?: string;
  sizeName?: string | null;
  status?: string;

  // Snapshot BEFORE any POS modifications (original product pricing)
  before: ItemPricingSnapshot;

  // Was item modified by POS operator?
  hadChangeFromPOS: boolean;

  // Snapshot AFTER POS modifications (current/final pricing)
  after: ItemPricingSnapshot;

  // Flat pricing fields (used by backend as fallback / legacy support)
  originalPrice?: number;
  currentPrice?: number;
  finalPrice?: number;
  hasActivePromotion?: boolean;
  overridePrice?: number;
  promotionType?: string | null;
  promotionValue?: number | null;
  totalBeforeDiscount?: number;
  discountAmount?: number;
  totalPrice?: number;
}

export interface DeliveryOptionRequest {
  name: string;
  description?: string;
  imageUrl?: string;
  price: number;
}

// Cart summary — matches backend POSCheckoutRequest.CartSummary
export interface CartSummary {
  businessId: string;
  businessName?: string;
  items: POSCheckoutItemRequest[];
  totalItems: number;
  totalQuantity: number;
  subtotalBeforeDiscount: number;
  subtotal: number;
  discountAmount: number;
  finalTotal: number;
}

export interface POSCheckoutAddressRequest {
  village: string;
  commune: string;
  district: string;
  province: string;
  streetNumber?: string;
  houseNumber?: string;
  note?: string;
  latitude?: number;
  longitude?: number;
}

// Order-level pricing audit trail — matches backend POSCheckoutRequest.PricingInfo
export interface PricingInfo {
  before: OrderPricingSnapshot;
  hadOrderLevelChangeFromPOS: boolean;
  after: OrderPricingSnapshot;
  orderLevelChangeReason?: string;
}

export interface PaymentInfo {
  paymentMethod: string;
  paymentStatus?: string;
}

// Top-level checkout request — matches backend POSCheckoutRequest exactly
export interface POSCheckoutRequest {
  businessId: string;
  customerId?: string;
  customerName?: string;
  customerPhone?: string;

  deliveryAddress: POSCheckoutAddressRequest;
  deliveryOption: DeliveryOptionRequest;
  cart: CartSummary;

  // Order-level pricing with before/after audit trail
  pricing: PricingInfo;

  payment: PaymentInfo;

  orderStatus?: string;
  customerNote?: string;
  businessNote?: string;
}

export interface POSCheckoutResponse {
  id: string;
  orderNumber: string;
  subtotal: number;
  discountAmount: number;
  deliveryFee: number;
  taxAmount: number;
  totalAmount: number;
  orderStatus: string;
  source: string;
  paymentMethod: string;
  paymentStatus: string;
  createdBy: string;
  createdAt: string;
  customerName?: string;
  customerPhone?: string;
}

File: menu-scanner-backend/src/main/java/com/emenu/features/order/dto/request/POSCheckoutRequest.java
package com.emenu.features.order.dto.request;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

/**
 * Unified POS checkout request
 *
 * Frontend → Backend mapping:
 * - deliveryAddress: full address object
 * - deliveryOption: full delivery option object (includes price)
 * - cart: CartSummary with items array and totals
 * - payment: PaymentInfo object
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class POSCheckoutRequest {

    @NotNull(message = "Business ID is required")
    private UUID businessId;

    // Customer info (optional for walkup orders)
    private UUID customerId;
    private String customerName;
    private String customerPhone;
    private String customerEmail;

    // Delivery address - use ID to fetch from database
    @NotNull(message = "Address ID is required")
    private UUID addressId;

    // Delivery option (full object with price, not just ID)
    @Valid
    private DeliveryOptionInfo deliveryOption;

    // Cart summary with all items and totals
    @Valid
    private CartSummary cart;

    // Pricing information with audit trail (order-level discounts)
    @Valid
    private PricingInfo pricing;

    // Payment information
    @Valid
    private PaymentInfo payment;

    // Order status
    private String orderStatus; // PENDING, CONFIRMED, COMPLETED, etc.

    // Notes for tracking
    private String customerNote;
    private String businessNote; // Includes discount reason and audit info

    // ─── Nested Classes ───
    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class DeliveryOptionInfo {
        private String name;
        private String description;
        private String imageUrl;
        private BigDecimal price;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class CartSummary {
        private UUID businessId;
        private String businessName;

        @NotEmpty(message = "Order must have at least one item")
        @Valid
        private List<POSCheckoutItemRequest> items;

        // Totals
        private Integer totalItems;
        private Integer totalQuantity;
        private BigDecimal subtotalBeforeDiscount;  // Sum of all original prices
        private BigDecimal subtotal;                // After product-level discounts
        private BigDecimal totalDiscount;           // Total discount from promotions
        private BigDecimal finalTotal;              // After order-level discount applied
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class PricingInfo {
        // Type of order-level discount (PERCENTAGE or FIXED_AMOUNT)
        private String discountType;

        // Reason for order-level change
        private String orderLevelChangeReason;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class PaymentInfo {
        @NotNull(message = "Payment method is required")
        private String paymentMethod;  // CASH

        private String paymentStatus;  // PAID, UNPAID, etc.
    }
}
